### PR TITLE
[quant][graph] Add JIT passes for dynamic quant multi uses of quant node

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37125 [quant][graph] Add JIT passes for dynamic quant multi uses of quant node**
* #37093 [quant][graph] Run dynamic quantization for specific ops
* #37045 [graph][quant] Enable accessing child/grandchild modules in forward
* #37014 [quant][graph] Add check for qconfig_dict key

Summary:
For dynamic quant we need to replicate the choose_qparams and quantize function in addition to replicating dequant.
RemoveRedundantQuantizeOps pass checks for the choose_qparams - quant - dequant pattern in the graph and removes it if the node following it cannot be quantized using dynamic quantization.

Test Plan:
python test_quantize_script.py test_dynamic_quant_multi_uses

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21283697](https://our.internmc.facebook.com/intern/diff/D21283697)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -2993,17 +3059,55 @@ void FoldQuantizedPrepackingOps(Module& module) {
   PrePackingOpsFolder(module, filter_fn, "quantized");
 }
 
+void RemoveRedundantQuantizeOps(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3062]: you can split this to a separate PR

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -2993,17 +3059,55 @@ void FoldQuantizedPrepackingOps(Module& module) {
...
+    Value* dequant_out = dequant_node->output();
+    TORCH_CHECK(
+        dequant_out->uses().size() == 1,
+        "Expect dequant output to have single use");
+    Node* user = dequant_out->uses()[0].user;
+    if (!nodeQuantizable(user, is_dynamic)) {
+      SubgraphRewriter rewriter;
+      rewriter.RegisterRewritePattern(
+          dynamic_quant_ops.pattern_string, dynamic_quant_replacement);
+      rewriter.runOnGraph(graph);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3091]: you can add a filter for this rewrite

File: torch/csrc/jit/passes/quantization.cpp (hunk 1 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:3081]: I feel this is still related to the change of replicating choose_qparams and quant. We shouldn't have a case of extra choose_qparams and quant nodes in the graph because it will affect perf. The test checks for this as well.
If you think splitting it helps in some other way then let me know.

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:3081]: it's a different pass though, you can still land them together even after you split. 
But I don't have a strong preference here.

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
...
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+    }
+  }
+  for (Node* n : quant_nodes_to_rewrite) {
+    Value* original_val = n->inputs()[0];
+    Value* quantized_val = n->output();
+    Value* dtype = n->inputs()[3];
+    TORCH_CHECK(original_val->uses().size() == 2);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:2742]: what is this check?

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
...
 
+void ReplicateChooseQParamsQuant(std::shared_ptr<Graph>& graph) {
+  std::stack<Block*> blocks_to_visit;
+  std::vector<Node*> quant_nodes_to_rewrite;
+  std::vector<Node*> choose_qparam_nodes_to_rewrite;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:2727]: I think it's better to check for the call sequence of "choose_qparams - quant - dequant" here directly

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:2742]: The original value should only have 2 uses i.e. choose_qparams and quantize_per_tensor. I

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:2727]: you can use subgraph matcher here as well

Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:2727]: At this point we have already called replicateDequant so the dequant nodes have been replicated. This way we can re-use that code from static and add this as a special case for dynamic

File: torch/csrc/jit/passes/quantization.cpp [hunk 5]
```diff
@@ -3011,16 +3078,48 @@ void FoldQuantizedPrepackingOps(Module& module) {
...
   ConstantPropagation(graph);
   ReplicateQuant(graph);
   ReplicateDeQuant(graph);
+  if (is_dynamic) {
+    ReplicateChooseQParamsQuant(graph);
+  }
   SwapDeQuant(graph);
   InsertPrepackUnpack(graph);
   ConstantPropagation(graph);
+  RemoveRedundantQuantizeOps(graph, is_dynamic);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3122]: can we do this immediately after ReplicateChooseQParamsQuant?

File: torch/csrc/jit/passes/quantization.cpp [hunk 6]
```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
...
+          insertChooseQParamsCall(graph.get(), original_val, user);
+      std::vector<Value*> quant_inputs = {original_val};
+      for (auto& out : choose_qparam->outputs()) {
+        quant_inputs.push_back(out);
+      }
+      quant_inputs.push_back(dtype);
+      Node* quant = graph->create(n->kind(), quant_inputs);
+      quant->output()->setDebugName(original_val->debugName() + ".quant");
+      graph->insertNode(quant);
+      user->replaceInputWith(quantized_val, quant->output());
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:2762]: I think there should be similar code when you insert choose_qpram - quant - dequant in the beginning, can you refactor this into a function?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:2762]: probably need some refactor of insertQuantDeQuantCall

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
please match the call sequence "choose_qparam - quant - dequant' as discussed in chat.

File: test/quantization/test_quantize_script.py [hunk 7]
```diff
@@ -1834,6 +1834,26 @@ def forward(self, x):
                    .check("quantized::linear_dynamic") \
                    .run(model.graph)
 
+    def test_dynamic_quant_multi_uses(self):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1837]: this is a test for finalize pass, so can be considered as a unit test.

File: torch/csrc/jit/passes/quantization.cpp [hunk 8]
```diff
@@ -2708,6 +2740,68 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
...
+    for (const Use& use : uses) {
+      auto* user = use.user;
+      WithInsertPoint ins(user);
+      auto quant_nodes = insertChooseQParamQuantDequant(
+          graph.get(), original_val, dtype, quant_node->kind());
+      user->replaceInputWith(dequant_out, std::get<2>(quant_nodes)->output());
+    }
+  }
+  for (const auto& n : nodes_to_rewrite) {
+    // dequantize
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:2791]: nit: can you unpack `n` here to `dequantize, quantize, choose_qparams` and then call remove on these things? similar for the next loop.

File: torch/csrc/jit/passes/quantization.cpp [hunk 9]
```diff
@@ -2708,6 +2740,68 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
...
+        %a_dequant = aten::dequantize(%a_quant)
+        return (%a_dequant) )");
+  const Graph& dynamic_quant_graph = *dynamic_quant_pattern.pattern_graph;
+  const auto& vmap = dynamic_quant_pattern.vmap;
+  Value* dequant_val = vmap.at("a_dequant");
+  Node* pattern_dequant = dequant_val->node();
+  Value* quant_val = vmap.at("a_quant");
+  Node* pattern_quant = quant_val->node();
+  Value* choose_qparam_val = vmap.at("a_scale");
+  Node* pattern_choose_qparam = choose_qparam_val->node();
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:2757]: nit: these can be moved to after the check for matches

File: torch/csrc/jit/passes/quantization.cpp [hunk 10]
```diff
@@ -3010,13 +3104,46 @@ void FoldQuantizedPrepackingOps(Module& module) {
...
 script::Module Finalize(script::Module& module, bool is_dynamic) {
   SwapFunctionalLinear(module);
   auto graph = module.get_method("forward").graph();
   Inline(*graph);
   ConstantPropagation(graph);
-  ReplicateQuant(graph);
-  ReplicateDeQuant(graph);
+  if (!is_dynamic) {
+    ReplicateQuant(graph);
+    ReplicateDeQuant(graph);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3142]: I feel ReplicateDequant is still useful in case the weight is used by multiple ops?

File: torch/csrc/jit/passes/quantization.cpp [hunk 11]
```diff
@@ -3010,13 +3104,46 @@ void FoldQuantizedPrepackingOps(Module& module) {
...
+  } else {
+    ReplicateChooseQParamsQuantDequant(graph);
+  }
+  RemoveRedundantQuantizeOps(graph, is_dynamic);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3146]: I think we just need to run these two passes before ReplicateDeQuant

File: torch/csrc/jit/passes/quantization.cpp [hunk 12]
```diff
@@ -1550,24 +1583,23 @@ void insertQuantDeQuantCall(
...
+      // Skip quant node and observer node (we need to keep
+      // observer nodes around since we need them to
+      // find the quantization parameters)
+      if (use.user != quant && use.user != observer &&
+          use.user != choose_qparams) {
+        uses.push_back(use);
+      }
     }
+    // in second pass, replace the original observed value with dequant output
+    insertDeQuantForAllUse(g, quant->output(), original_val, uses);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1601]: we don't need to insert dequant for all use here anymore since we have replicate dequant pass in the end.

File: torch/csrc/jit/passes/quantization.cpp [hunk 13]
```diff
@@ -1486,16 +1500,44 @@ void insertDeQuantCall(
...
-        ->setType(original_val->type());
+    Node* dequant = insertDeQuant(graph, quantized_val, original_val, i);
     user->replaceInputWith(original_val, dequant->output());
-    graph->insertNode(dequant);
   }
 }
 
+using DynamicQuantOps = std::tuple<Node*, Node*, Node*>;
+
+DynamicQuantOps insertChooseQParamQuantDequant(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1510]: can you break this down further to insertChooseQParams and insertQuant?

File: test/quantization/test_quantize_script.py (hunk 7 above)
Comment (supriyar) [reply test/quantization/test_quantize_script.py:1837]: Yeah I'll refactor the dynamic quant tests in a follow-up PR.

File: torch/csrc/jit/passes/quantization.cpp [hunk 14]
```diff
@@ -1515,32 +1557,23 @@ void insertQuantDeQuantCall(
...
+    v->replaceAllUsesWith(original_val);
+    std::vector<Use> uses = original_val->uses();
+    for (const auto& use : uses) {
+      auto* user = use.user;
+      if (user != std::get<1>(quant_nodes) && user != observer &&
+          user != std::get<0>(quant_nodes)) {
+        user->replaceInputWith(
+            original_val, std::get<2>(quant_nodes)->output());
+      }
+    }
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1576]: I think this code can be shared by two branches?

File: torch/csrc/jit/passes/quantization.cpp [hunk 15]
```diff
@@ -1515,32 +1557,23 @@ void insertQuantDeQuantCall(
   } else {
     quantize_func = "quantize_per_tensor";
   }
+  Value* original_val = observer->input(1);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1560]: could you add comments for `original_val` and `v`, it's starting to get a bit confusing

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
RemoveRedundantQuantizeOps looks good, but insertQuantDeQuant might need more refactor

File: torch/csrc/jit/passes/quantization.cpp (hunk 13 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:1510]: since insertQuant can be reused by another branch

File: torch/csrc/jit/passes/quantization.cpp [hunk 16]
```diff
@@ -1486,16 +1500,44 @@ void insertDeQuantCall(
...
+  }
+  quant_inputs.push_back(dtype);
+  Node* quant = graph->create(quant_kind, quant_inputs);
+  quant->output()->setDebugName(original_val->debugName() + ".quant");
+  graph->insertNode(quant);
+  Node* dequant = insertDeQuant(graph, quant->output(), original_val);
+  return std::make_tuple(choose_qparams, quant, dequant);
+}
+
 void insertQuantDeQuantCall(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1541]: also this function should probably be renamed now, maybe things like insertNumericsChangingOps

File: torch/csrc/jit/passes/quantization.cpp (hunk 12 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:1601]: Sounds good. I was wondering if I should remove this

File: torch/csrc/jit/passes/quantization.cpp (hunk 11 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:3146]: Sounds good

File: test/quantization/test_quantize_script.py [hunk 17]
```diff
@@ -762,15 +762,15 @@ def forward(self, x, w0, w1, w2):
...
-
-        # we just check we have one dequant on every op input, even input
-        # is sharded as multi uses
-        FileCheck().check_count("aten::dequantize", 9, exactly=True) \
+        m = wrap_cpp_module(torch._C._jit_pass_insert_quant_dequant(m._c, "forward", False))
+        # We replicate dequant in finalize call
+        m = wrap_cpp_module(torch._C._jit_pass_quant_finalize(m._c))
+        # check that quantized ops are inserted correctly.
+        FileCheck().check_count("quantized::conv2d_prepack", 3, exactly=True) \
                    .run(str(get_forward_graph(m._c)))
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:774]: this test can be removed actually, we already have tests for replicate_dequant

File: torch/csrc/jit/passes/quantization.cpp [hunk 18]
```diff
@@ -1475,7 +1475,32 @@ void InsertObserversHelper::propagateObservedProperty(
   }
 }
 
-void insertDeQuantCall(
+Node* insertQuant(
+    Graph* graph,
+    std::vector<Value*> inputs,
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1480]: please fix lint

File: torch/csrc/jit/passes/quantization.cpp [hunk 19]
```diff
@@ -1486,17 +1511,44 @@ void insertDeQuantCall(
...
+  Node* choose_qparams = graph->create(
+      at::Symbol::aten(choose_qparams_func),
+      {original_val, reduce_range},
+      /* num_outputs = */ 2);
+  choose_qparams->output(0)->setDebugName(original_val->debugName() + ".scale");
+  choose_qparams->output(0)->setType(FloatType::get());
+  choose_qparams->output(1)->setDebugName(
+      original_val->debugName() + ".zero_point");
+  choose_qparams->output(1)->setType(IntType::get());
+  graph->insertNode(choose_qparams);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1538]: to be more consistent, could you add a function for insertChooseQParams as well?

File: torch/csrc/jit/passes/quantization.cpp [hunk 20]
```diff
@@ -1505,69 +1557,43 @@ void insertQuantDeQuantCall(
...
-  for (const auto& use : original_val->uses()) {
-    // Skip quant node and observer node (we need to keep
-    // observer nodes around since we need them to
-    // find the quantization parameters)
-    if (use.user != quant && use.user != observer &&
-        use.user != choose_qparams) {
-      uses.push_back(use);
+  observer_out->replaceAllUsesWith(original_val);
+  std::vector<Use> uses = original_val->uses();
+  for (const auto& use : uses) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1591]: nit: you can merge these two lines now

File: torch/csrc/jit/passes/quantization.cpp [hunk 21]
```diff
@@ -3010,11 +3100,43 @@ void FoldQuantizedPrepackingOps(Module& module) {
...
+}
+
 script::Module Finalize(script::Module& module, bool is_dynamic) {
   SwapFunctionalLinear(module);
   auto graph = module.get_method("forward").graph();
   Inline(*graph);
   ConstantPropagation(graph);
+  if (is_dynamic) {
+    ReplicateChooseQParamsQuantDequant(graph);
+    RemoveRedundantQuantizeOps(graph, is_dynamic);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3138]: `is_dynamic` seems to be redundant here?

File: torch/csrc/jit/passes/quantization.cpp [hunk 22]
```diff
@@ -3010,11 +3100,43 @@ void FoldQuantizedPrepackingOps(Module& module) {
...
+  auto filter = [&](const Match& match,
+                    const std::unordered_map<std::string, Value*>& vmap) {
+    const auto& match_vmap = match.values_map;
+    auto dequant_node = match_vmap.at(vmap.at("a_dequant"))->node();
+    Value* dequant_out = dequant_node->output();
+    TORCH_CHECK(
+        dequant_out->uses().size() == 1,
+        "Expect dequant output to have single use");
+    Node* user = dequant_out->uses()[0].user;
+    return !nodeQuantizable(user, is_dynamic);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:3124]: should we just use is_dynamic=True here?

File: torch/csrc/jit/passes/quantization.cpp [hunk 23]
```diff
@@ -1486,17 +1511,44 @@ void insertDeQuantCall(
...
+    quant_inputs.push_back(out);
+  }
+  quant_inputs.push_back(dtype);
+  Node* quant = insertQuant(
+      graph, quant_inputs, quant_kind, original_val->debugName() + ".quant");
+  Node* dequant = insertDeQuant(graph, quant->output(), original_val);
+  return std::make_tuple(choose_qparams, quant, dequant);
+}
+
+void insertQuantizationOps(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1551]: Thanks! this function looks pretty clear now

**Review by jerryzh168 — APPROVED** [review]
Had a few inline comments, please address them before landing.

File: torch/csrc/jit/passes/quantization.cpp (hunk 21 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:3138]: Yes it is, I was just trying to separate out the code paths. But can remove it if you prefer that.
